man/roff.7.in
'\" t -- preprocess: tbl(1)
.ig
@ roff.7

Copyright (c) 2014 - 2015 Steffen (Daode) Nurpmeso <sdaoden@users.sf.net>.

Copyright (C) 2000 - 2004, 2006 - 2008
Free Software Foundation, Inc.
written by Bernd Warken <bwarken@mayn.de>
maintained by Werner Lemberg <wl@gnu.org>

Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.1 or
any later version published by the Free Software Foundation; with the
Invariant Sections being this .ig-section and AUTHORS, with no
Front-Cover Texts, and with no Back-Cover Texts.

You should have received a copy of the Free Documentation License
as part of the file COPYING; also located in the main directory of the
source package of this program.
..
.
.\"
.\" Local macros
.
.de Esc
.  ds @1 \\$1
.  shift
.  nop \f[B]\[rs]\\*[@1]\f[]\\$*
.  rm @1
..
.
.
.de QuotedChar
.  ds @1 \\$1
.  shift
.  nop `\f[B]\\*[@1]\f[]'\\$*
.  rm @1
..
.
.
.\"
.\" Title
.\"
.
.TH ROFF @MAN7EXT@ "@MDATE@" "@T_ROFF@ v@VERSION@"
.SH NAME
roff \- concepts and history of roff typesetting
.
.
.\"
.SH DESCRIPTION
.\"
.
.I roff
is the general name for a set of text formatting programs, known under
names like
.BR troff ,
.BR nroff ,
.BR ditroff
etc.
.
A
.I roff
system consists of an extensible text formatting language and a set of
programs for printing and converting to other text formats.
.
Unix-like operating systems distribute a
.I roff
system as a core package.
.
.P
The most common
.I roff
system today is the free software implementation \f[CR]GNU\f[]
.IR roff.
@T_ROFF@ is an deriviative.
Both implement the look-and-feel and functionality of its ancestors, with
many extensions.
.
.P
The ancestry of
.I roff
is described in section
.BR HISTORY .
.
In this document, the term
.I roff
always refers to the general class of roff programs, not to the
.B roff
command provided in early UNIX systems.
.
.P
In spite of its age,
.I roff
is in wide use today, for example, the manual pages on UNIX systems
.RI ( man\~pages\/ ),
many software books, system documentation, standards, and corporate
documents are written in roff.
.
The
.I roff
output for text devices is still unmatched, and its graphical output
has the same quality as other free type-setting programs and is better
than some of the commercial systems.
.
.P
.I roff
is used to format UNIX
.IR "manual pages" ,
(or
.IR "man pages" ),
the standard documentation system on many UNIX-derived operating systems.
.
.P
This document describes the history of the development of the
.IR "roff system" ;
some usage aspects common to all
.I roff
versions, details on the
.I roff
pipeline, which is usually hidden behind front-ends like
.BR @L_ROFF@ (@MAN1EXT@);
a general overview of the formatting language; some tips for editing
.I roff
files; and many pointers to further readings.
.
.
.\"
.SH "HISTORY"
.\"
.
Document formatting by computer dates back to the 1960s.
.
The
.I roff
system itself is intimately connected to the Unix operating system, but its
roots go back to the earlier operating systems CTSS and Multics.
.
.
.\"
.SS "The Predecessor RUNOFF"
.\"
.
.BR roff 's
ancestor
.B RUNOFF
was written in the MAD language by
.I Jerry Saltzer
for the
.IR "Compatible Time Sharing System (CTSS)" ,
a project of the Massachusetts Institute of Technology (MIT), in 1963 and
1964 \[en] note that CTSS commands were all uppercase.
.
.P
In 1965, MIT's Project MAC teamed with Bell Telephone Laboratories (BTL)
and General Electric to begin the
.UR http://\:www.multicians.org
.I Multics
system
.UE .
.
A command called
.B runoff
was written for Multics in the late 60s in the BCPL language, by
.IR "Bob Morris" ,
.IR "Doug McIlroy" ,
and other members of the Multics team.
.
.P
Like its CTSS ancestor, Multics
.B runoff
formatted an input file consisting of text and command lines; commands began
with a period and were two letters.
.
Output from these commands was to terminal devices such as IBM Selectric
terminals.
.
Multics
.B runoff
had additional features added, such as the ability to do two-pass
formatting; it became the main format for Multics documentation and text
processing.
.
.P
BCPL and
.B runoff
were ported to the GCOS system at Bell Labs when BTL left the development of
Multics.
.
.
.\"
.SS "The Classical nroff/troff System"
.\"
.
At BTL, there was a need to drive the
.I Graphic Systems CAT
typesetter, a graphical output device from a PDP-11 computer running Unix.
.
As
.B runoff
was too limited for this task it was further developed into a more
powerful text formatting system by
.IR "Joseph F.\& Ossanna" ,
who already programmed several runoff ports.
.
.P
The name
.I runoff
was shortened to
.IR roff .
.
The greatly enlarged language of Ossanna's version already included all
elements of a full
.IR "roff system" .
.
All modern
.I roff
systems try to implement compatibility to this system.
.
So Joe Ossanna can be called the father of all
.I roff
systems.
.
.P
This first
.I roff system
had three formatter programs.
.
.TP
.B troff
.RI ( "typesetter roff\/" )
generated a graphical output for the
.I CAT
typesetter as its only device.
.
.TP
.B nroff
produced text output suitable for terminals and line printers.
.
.TP
.B roff
was the reimplementation of the former
.B runoff
program with its limited features; this program was abandoned in later
versions.
.
Today, the name
.I roff
is used to refer to a
.I troff/\:nroff
sytem as a whole.
.
.P
Ossanna's first version was written in the PDP-11 assembly language and
released in 1973.
.
.I Brian Kernighan
joined the
.I roff
development by rewriting it in the C\~programming language.
.
The C\~version was released in 1975.
.
.P
The syntax of the formatting language of the
.BR nroff /\: troff
programs was documented in the famous
.IR "Troff User's Manual [CSTR\~#54]" ,
first published in 1976, with further revisions up to 1992 by Brian
Kernighan.
.
This document is the specification of the
.IR "classical troff" .
.
All later
.I roff
systems tried to establish compatibility with this specification.
.
.P
After Ossanna's death in 1977, Kernighan went on with developing
.IR troff .
.
In the late 1970s, Kernighan equipped
.I troff
with a general interface to support more devices, the intermediate
output format, and the postprocessor system.
.
This completed the structure of a
.I "roff system"
as it is still in use today; see section
.BR "USING ROFF" .
.
In 1979, these novelties were described in the paper
.IR "[CSTR\~#97]" .
.
This new
.I troff
version is the basis for all existing newer troff systems, including
.IR @L_ROFF@ .
.
On some systems, this
.I device independent troff
got a binary of its own, called
.BR ditroff (@MAN7EXT@).
.
All modern
.B troff
programs already provide the full
.B ditroff
capabilities automatically.
.
.
.\"
.SS "Availability"
.\"
.
The source code of both the ancient Unix and classical
.I troff
weren't available for two decades.
.
Meanwhile, it is accessible again (on-line) for non-commercial use,
cf.\& section
.BR "SEE ALSO" .
.
.
.\"
.SS "Free roff"
.\"
.
The most important free
.I roff
project was the \f[CR]GNU\f[] implementation of
.IR troff ,
written from scratch by
.I James Clark
and put under the
.UR http://\:www.gnu.org/\:copyleft
GNU Public License
.UE .
.
It was called
.I groff
(\f[CR]GNU\f[]
.IR roff ).
.
Further forks exists.
See
.BR @L_ROFF@ (@MAN1EXT@)
for an overview.
.
It is compatible to the classical
.IR troff ,
but many extensions were added.
.
It is the first
.I roff
system that is available on almost all operating systems \[en] and it is
free.
.
.
.\"
.SH "USING ROFF"
.\"
.
Most people won't even notice that they are actually using
.IR roff .
.
When you read a system manual page (man page)
.I roff
is working in the background.
.
.P
Some
.I roff
implementations provide wrapper programs that make it easy to use the
.I roff
system on the shell command line.
.
For example, the \f[CR]GNU\f[]
.I roff
implementation
.BR groff (@MAN1EXT@)
provides command line options to avoid the long command pipes of
classical
.IR troff .
.
.
.\"
.SS "The roff Pipe"
.\"
.
Each
.I roff
system consists of preprocessors,
.I roff
formatter programs, and a set of device postprocessors.
.
This concept makes heavy use of the
.I piping
mechanism, that is, a series of programs is called one after the other,
where the output of each program in the queue is taken as the input
for the next program.
.
.RS
.P
cat \f[I]file\f[P] \
| .\|.\|. \
| \f[I]preproc\f[P] \
| .\|.\|. \
| troff \f[I]options\f[P] \
| \f[I]postproc\f[P]
.RE
.
.P
The preprocessors generate
.I roff
code that is fed into a
.I roff
formatter (e.g.\&
.BR troff ),
which in turn generates
.I intermediate output
that is fed into a device postprocessor program for printing or final
output.
.
.P
All of these parts use programming languages of their own; each
language is totally unrelated to the other parts.
.
Moreover,
.I roff
macro packages that were tailored for special purposes can be
included.
.
.P
Most
.I roff
documents use the macros of some package, intermixed with code for one
or more preprocessors, spiced with some elements from the plain
.I roff
language.
.
The full power of the
.I roff
formatting language is seldom needed by users; only programmers of
macro packages need to know about the gory details.
.
.
.
.\"
.SS "Preprocessors"
.\"
.
A
.I roff
preprocessor is any program that generates output that syntactically
obeys the rules of the
.I roff
formatting language.
.
Each preprocessor defines a language of its own that is translated
into
.I roff
code when run through the preprocessor program.
.
Parts written in these languages may be included within a
.I roff
document; they are identified by special
.I roff
requests or macros.
.
Each document that is enhanced by preprocessor code must be run
through all corresponding preprocessors before it is fed into the
actual
.I roff
formatter program, for the formatter just ignores all alien code.
.
The preprocessor programs extract and transform only the document
parts that are determined for them.
.
.P
There are a lot of free and commercial
.I roff
preprocessors.
.
Some of them aren't available on each system, but there is a small
set of preprocessors that are considered as an integral part of each
.I roff
system.
.
The classical preprocessors are
.
.RS
.TS
tab (@);
lb l.
tbl@for tables.
eqn@for mathematical formul\[ae].
pic@for drawing diagrams.
refer@for bibliographic references.
soelim@for including macro files from standard locations.
chem@for drawing chemical formul\[ae].
.TE
.RE
.
.P
Other known preprocessors that are not available on all systems
include
.
.RS
.TS
tab (@);
lb l.
grap@for constructing graphical elements.
grn@for including \fBgremlin\fR(1) pictures.
.TE
.RE
.
.\"
.SS "Formatter Programs"
.\"
.
A
.I roff formatter
is a program that parses documents written in the
.I roff
formatting language or uses some of the
.I roff
macro packages.
.
It generates
.IR "intermediate output" ,
which is intended to be fed into a single device postprocessor that
must be specified by a command-line option to the formatter program.
.
The documents must have been run through all necessary preprocessors
before.
.
.P
The output produced by a
.I roff
formatter is represented in yet another language, the
.IR "intermediate output format"
or
.IR "troff output" .
.
This language was first specified in
.IR "[CSTR\~#97]" ;
its @T_ROFF@ extension is documented in
.BR @L_ROFF@_out (@MAN5EXT@).
.
The intermediate output language is a kind of assembly language
compared to the high-level
.I roff
language.
.
The generated intermediate output is optimized for a special device,
but the language is the same for every device.
.
.P
The
.I roff
formatter is the heart of the
.I roff
system.
.
The traditional
.I roff
had two formatters,
.B nroff
for text devices and
.B troff
for graphical devices.
.
.P
Often, the name
.I troff
is used as a general term to refer to both formatters.
.
.
.\"
.SS "Devices and Postprocessors"
.\"
.
Devices are hardware interfaces like printers, text or graphical
terminals, etc., or software interfaces such as a conversion into a
different text or graphical format.
.
.P
A
.I roff
postprocessor is a program that transforms
.I troff
output into a form suitable for a special device.
.
The
.I roff
postprocessors are like device drivers for the output target.
.
.P
For each device there is a postprocessor program that fits the device
optimally.
.
The postprocessor parses the generated intermediate output and
generates device-specific code that is sent directly to the device.
.
.P
The names of the devices and the postprocessor programs are not fixed
because they greatly depend on the software and hardware abilities of
the actual computer.
.
For example, the classical devices mentioned in
.I [CSTR\~#54]
have greatly changed since the classical times.
.
The old hardware doesn't exist any longer and the old graphical
conversions were quite imprecise when compared to their modern
counterparts.
.
.P
For example, the Postscript device
.I post
in classical
.I troff
had a resolution of 720 units per inch, while
.IR @T_ROFF@ 's
.I ps
device has 72000, a refinement of factor 100.
.
.P
Today the operating systems provide device drivers for most
printer-like hardware, so it isn't necessary to write a special
hardware postprocessor for each printer.
.
.
.\"
.SH "ROFF PROGRAMMING"
.\"
.
Documents using
.I roff

NEWS
S - r o f f  N e w s
====================

The complete changelog of commits in between two versions can be
inspected by using the git(1) "log" command as shown below, where "OLD"
and "NEW" are the two versions to be compared, e.g., v2.0 and v2.1:

  # All commits:
  $ git log --reverse --topo-order --abbrev-commit OLD..NEW
  # Only topic branch headers (--no-merges for content commits only):
  $ git log --oneline --reverse --topo-order --merges OLD..NEW
  # Same, but truly accessible:
  $ git log --oneline --reverse --topo-order --merges --parents OLD..NEW |
    while read c1 c2 c3 c4 c5 c6; do
      printf "%-24s: \$ git log --oneline --no-merges ${c1} ^${c2}\n" "${c6}";
    done

v2.0.0, 20??-??-??
------------------



s-roff, 2014-08-19
------------------

S-roff has been forked from GNU troff, revision 1.19.2-574-gecbf4f1
(from the official GNU troff git(1) repository; the last commit that
is still licensed under GPLv2).

This file is stripped down for major releases: to see NEWS for older
releases, use the git(1) revision control system on this file.  For
changes before the fork of S-roff, please look at all the ChangeLog
files that still exist for tag "s-roff" [ecbf4f1])

# s-ts-mode


src/roff/roff.cpp
/*@ A front end for S-roff.
 *
 * Copyright (c) 2014 - 2015 Steffen (Daode) Nurpmeso <sdaoden@users.sf.net>.
 *
 * Copyright (C) 1989 - 2008
 *    Free Software Foundation, Inc.
 *      Written by James Clark (jjc@jclark.com)
 *
 * This is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2, or (at your option) any later
 * version.
 *
 * This is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with groff; see the file COPYING.  If not, write to the Free Software
 * Foundation, 51 Franklin St - Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "config.h"
#include "roff-config.h"

#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>

#include "cset.h"
#include "defs.h"
#include "device.h"
#include "errarg.h"
#include "error.h"
#include "font.h"
#include "lib.h"
#include "nonposix.h"
#include "pipeline.h"
#include "stringclass.h"

// The number of commands must be in sync with MAX_COMMANDS in pipeline.h
enum {
  IDX_PRECONV,
  IDX_SOELIM,
  IDX_REFER,
  IDX_GRAP,
  IDX_PIC,
  IDX_TBL,
  IDX_GRN,
  IDX_EQN,
  IDX_TROFF,
  IDX_POST,
  IDX_SPOOL,
  IDX_NCOMMANDS
};
CTA(IDX_NCOMMANDS <= MAX_COMMANDS);

class possible_command
{
  char *name;
  string args;
  char **argv;

  void build_argv();

public:
  possible_command();
  ~possible_command();
  void clear_name();
  void set_name(const char *);
  void set_name(const char *, const char *);
  const char *get_name();
  void append_arg(const char *, const char * = 0);
  void insert_arg(const char *);
  void insert_args(string s);
  void clear_args();
  char **get_argv();
  void print(int is_last, FILE *fp);
};

int lflag = 0;
char *spooler = 0;
char *postdriver = 0;
char *predriver = 0;

possible_command commands[IDX_NCOMMANDS];

int run_commands(int no_pipe);
void print_commands(FILE *);
void append_arg_to_string(const char *arg, string &str);
void handle_unknown_desc_command(const char *command, const char *arg,
				 const char *filename, int lineno);
const char *xbasename(const char *); // FIXME lib.h??

void usage(FILE *stream);
void help();

int main(int argc, char **argv)
{
  program_name = argv[0];
  static char stderr_buf[BUFSIZ];
  setbuf(stderr, stderr_buf);
  string Pargs, Largs, Fargs;
  int Kflag = 0;
  int vflag = 0;
  int Vflag = 0;
  int zflag = 0;
  int iflag = 0;
  int oflag = 0;
  int safer_flag = 1;
  int is_xhtml = 0;
  int eflag = 0;
  int opt;
  const char *command_prefix = getenv(U_ROFF_COMMAND_PREFIX);
  const char *encoding = getenv(U_ROFF_ENCODING);
  if (!command_prefix)
    command_prefix = PROG_PREFIX;
  commands[IDX_TROFF].set_name(command_prefix, "troff");
  static const struct option long_options[] = {
    { "help", no_argument, 0, 'h' },
    { "version", no_argument, 0, 'v' },
    { NULL, 0, 0, 0 }
  };
  while ((opt = getopt_long(
		  argc, argv,
		  "abcCd:D:eEf:F:gGhiI:lkK:L:m:M:n:No:pP:r:RsStT:UvVw:W:XzZ",
		  long_options, NULL))
	 != EOF) {
    char buf[3];
    buf[0] = '-';
    buf[1] = opt;
    buf[2] = '\0';
    switch (opt) {
    case 'i':
      iflag = 1;
      break;
    case 'I':
      commands[IDX_SOELIM].set_name(command_prefix, "soelim");
      commands[IDX_SOELIM].append_arg(buf, optarg);
      // .psbb may need to search for files
      commands[IDX_TROFF].append_arg(buf, optarg);
      // \X'ps:import' may need to search for files
      Pargs += buf;
      Pargs += optarg;
      Pargs += '\0';
      break;
    case 'D':
      commands[IDX_PRECONV].set_name(command_prefix, "preconv");
      commands[IDX_PRECONV].append_arg("-D", optarg);
      break;
    case 'K':
      commands[IDX_PRECONV].append_arg("-e", optarg);
      Kflag = 1;
      // fall through
    case 'k':
      commands[IDX_PRECONV].set_name(command_prefix, "preconv");
      break;
    case 't':
      commands[IDX_TBL].set_name(command_prefix, "tbl");
      break;
    case 'p':
      commands[IDX_PIC].set_name(command_prefix, "pic");
      break;
    case 'g':
      commands[IDX_GRN].set_name(command_prefix, "grn");
      break;
    case 'G':
      commands[IDX_GRAP].set_name(command_prefix, "grap");
      break;
    case 'e':
      eflag = 1;
      commands[IDX_EQN].set_name(command_prefix, "eqn");
      break;
    case 's':
      commands[IDX_SOELIM].set_name(command_prefix, "soelim");
      break;
    case 'R':
      commands[IDX_REFER].set_name(command_prefix, "refer");
      break;
    case 'z':
    case 'a':
      commands[IDX_TROFF].append_arg(buf);
      // fall through
    case 'Z':
      zflag++;
      break;
    case 'l':
      lflag++;
      break;
    case 'V':
      Vflag++;
      break;
    case 'v':
      vflag = 1;
      printf(L_ROFF " v" VERSION);
      printf(L_ROFF_COPYRIGHT_PRINTOUT);
      printf("\ncalled subprograms:\n\n");
      fflush(stdout);
      commands[IDX_POST].append_arg(buf);
      // FALLTHRU
    case 'C':
      commands[IDX_SOELIM].append_arg(buf);
      commands[IDX_REFER].append_arg(buf);
      commands[IDX_PIC].append_arg(buf);
      commands[IDX_GRAP].append_arg(buf);
      commands[IDX_TBL].append_arg(buf);
      commands[IDX_GRN].append_arg(buf);
      commands[IDX_EQN].append_arg(buf);
      commands[IDX_TROFF].append_arg(buf);
      break;
    case 'N':
      commands[IDX_EQN].append_arg(buf);
      break;
    case 'h':
      help();
      break;
    case 'E':
    case 'b':
      commands[IDX_TROFF].append_arg(buf);
      break;
    case 'c':
      commands[IDX_TROFF].append_arg(buf);
      break;
    case 'S':
      safer_flag = 1;
      break;
    case 'U':
      safer_flag = 0;
      break;
    case 'T':
      if (strcmp(optarg, "xhtml") == 0) {
	// force soelim to aid the html preprocessor
	commands[IDX_SOELIM].set_name(command_prefix, "soelim");
	Pargs += "-x";
	Pargs += '\0';
	Pargs += 'x';
	Pargs += '\0';
	is_xhtml = 1;
	device = "html";
	break;
      }
      if (strcmp(optarg, "html") == 0)
	// force soelim to aid the html preprocessor
	commands[IDX_SOELIM].set_name(command_prefix, "soelim");
      device = optarg;
      break;
    case 'F':
      font::command_line_font_dir(optarg);
      if (Fargs.length() > 0) {
	Fargs += PATH_SEP_CHAR;
	Fargs += optarg;
      }
      else
	Fargs = optarg;
      break;
    case 'o':
      oflag = 1;
    case 'f':
    case 'm':
    case 'r':
    case 'd':
    case 'n':
    case 'w':
    case 'W':
      commands[IDX_TROFF].append_arg(buf, optarg);
      break;
    case 'M':
      commands[IDX_EQN].append_arg(buf, optarg);
      commands[IDX_GRAP].append_arg(buf, optarg);
      commands[IDX_GRN].append_arg(buf, optarg);
      commands[IDX_TROFF].append_arg(buf, optarg);
      break;
    case 'P':
      Pargs += optarg;
      Pargs += '\0';
      break;
    case 'L':
      append_arg_to_string(optarg, Largs);
      break;
    case '?':
      usage(stderr);
      exit(1);
      break;
    default:
      assert(0);
      break;
    }
  }
  if (encoding) {
    commands[IDX_PRECONV].set_name(command_prefix, "preconv");
    if (!Kflag && *encoding)
      commands[IDX_PRECONV].append_arg("-e", encoding);
  }
  if (!safer_flag) {
    commands[IDX_TROFF].insert_arg("-U");
    commands[IDX_PIC].append_arg("-U");
  }
  font::set_unknown_desc_command_handler(handle_unknown_desc_command);
  if (!font::load_desc())
    fatal("invalid device `%1'", device);
  if (!postdriver)
    fatal("no `postpro' command in DESC file for device `%1'", device);
  if (predriver && !zflag) {
    commands[IDX_TROFF].insert_arg(commands[IDX_TROFF].get_name());
    commands[IDX_TROFF].set_name(predriver);
    // pass the device arguments to the predrivers as well
    commands[IDX_TROFF].insert_args(Pargs);
    if (eflag && is_xhtml)
      commands[IDX_TROFF].insert_arg("-e");
    if (vflag)
      commands[IDX_TROFF].insert_arg("-v");
  }
  const char *real_driver = 0;
  if (postdriver)
    commands[IDX_POST].set_name(postdriver);
  const char *p = Pargs.contents();
  const char *end = p + Pargs.length();
  while (p < end) {
    commands[IDX_POST].append_arg(p);
    p = strchr(p, '\0') + 1;
  }
  if (lflag && !vflag && spooler) {
    commands[IDX_SPOOL].set_name(BSHELL);
    commands[IDX_SPOOL].append_arg(BSHELL_DASH_C);
    Largs += '\0';
    Largs = spooler + Largs;
    commands[IDX_SPOOL].append_arg(Largs.contents());
  }
  if (zflag) {
    commands[IDX_POST].set_name(0);
    commands[IDX_SPOOL].set_name(0);
  }
  commands[IDX_TROFF].append_arg("-T", device);
  if (strcmp(device, "html") == 0) {
    if (is_xhtml) {
      if (oflag)
	fatal("`-o' option is invalid with device `xhtml'");
      if (zflag)
	commands[IDX_EQN].append_arg("-Tmathml:xhtml");
      else if (eflag)
	commands[IDX_EQN].clear_name();
    }
    else {
      if (oflag)
	fatal("`-o' option is invalid with device `html'");
      // html renders equations as images via ps
      commands[IDX_EQN].append_arg("-Tps:html");
    }
  }
  else
    commands[IDX_EQN].append_arg("-T", device);

  commands[IDX_GRN].append_arg("-T", device);

  int first_index;
  for (first_index = 0; first_index < IDX_TROFF; ++first_index)
    if (commands[first_index].get_name() != 0)
      break;
  if (optind < argc) {
    if (argv[optind][0] == '-' && argv[optind][1] != '\0')
      commands[first_index].append_arg("--");
    for (int i = optind; i < argc; i++)
      commands[first_index].append_arg(argv[i]);
    if (iflag)
      commands[first_index].append_arg("-");
  }
  if (Fargs.length() > 0) {
    string e = U_ROFF_FONT_PATH;
    e += '=';
    e += Fargs;
    char *fontpath = getenv(U_ROFF_FONT_PATH);
    if (fontpath && *fontpath) {
      e += PATH_SEP_CHAR;
      e += fontpath;
    }
    e += '\0';
    if (putenv(strsave(e.contents())))
      fatal("putenv failed");
  }
  {
    // we save the original path in GROFF_PATH__ and put it into the
    // environment -- troff will pick it up later.
    char *path = getenv("PATH");
    string e = U_ROFF_PATH__;
    e += '=';
    if (path && *path)
      e += path;
    e += '\0';
    if (putenv(strsave(e.contents())))
      fatal("putenv failed");
    char *binpath = getenv(U_ROFF_BIN_PATH);
    string f = "PATH";
    f += '=';
    if (binpath && *binpath)
      f += binpath;
    else
      f += BINPATH;
    if (path && *path) {
      f += PATH_SEP_CHAR;
      f += path;
    }
    f += '\0';
    if (putenv(strsave(f.contents())))
      fatal("putenv failed");
  }
  if (Vflag)
    print_commands(Vflag == 1 ? stdout : stderr);
  if (Vflag == 1)
    exit(0);
  return run_commands(vflag);
}

const char *xbasename(const char *s) // TODO -> library
{
  if (!s)
    return 0;
  // DIR_SEPS[] are possible directory separator characters, see nonposix.h
  // We want the rightmost separator of all possible ones.
  // Example: d:/foo\\bar.
  const char *p = strrchr(s, DIR_SEPS[0]), *p1;
  const char *sep = &DIR_SEPS[1];

  while (*sep)
    {
      p1 = strrchr(s, *sep);
      if (p1 && (!p || p1 > p))
	p = p1;
      sep++;
    }
  return p ? p + 1 : s;
}

void handle_unknown_desc_command(const char *command, const char *arg,
				 const char *filename, int lineno)
{
  if (strcmp(command, "print") == 0) {
    if (arg == 0)
      error_with_file_and_line(filename, lineno,
			       "`print' command requires an argument");
    else
      spooler = strsave(arg);
  }
  if (strcmp(command, "prepro") == 0) {
    if (arg == 0)
      error_with_file_and_line(filename, lineno,
			       "`prepro' command requires an argument");
    else {
      for (const char *p = arg; *p; p++)
	if (csspace(*p)) {
	  error_with_file_and_line(filename, lineno,
				   "invalid `prepro' argument `%1'"
				   ": program name required", arg);
	  return;
	}
      predriver = strsave(arg);
    }
  }
  if (strcmp(command, "postpro") == 0) {
    if (arg == 0)
      error_with_file_and_line(filename, lineno,
			       "`postpro' command requires an argument");
    else {
      for (const char *p = arg; *p; p++)
	if (csspace(*p)) {
	  error_with_file_and_line(filename, lineno,
				   "invalid `postpro' argument `%1'"
				   ": program name required", arg);
	  return;
	}
      postdriver = strsave(arg);
    }
  }
}

void print_commands(FILE *fp)
{
  int last;
  for (last = IDX_SPOOL; last >= 0; last--)
    if (commands[last].get_name() != 0)
      break;
  for (int i = 0; i <= last; i++)
    if (commands[i].get_name() != 0)
      commands[i].print(i == last, fp);
}

// Run the commands. Return the code with which to exit.

int run_commands(int no_pipe)
{
  char **v[IDX_NCOMMANDS];
  int j = 0;
  for (int i = 0; i < IDX_NCOMMANDS; i++)
    if (commands[i].get_name() != 0)
      v[j++] = commands[i].get_argv();
  return run_pipeline(j, v, no_pipe);
}

possible_command::possible_command()
: name(0), argv(0)
{
}

possible_command::~possible_command()
{
  a_delete name;
  a_delete argv;
}

void possible_command::set_name(const char *s)
{
  a_delete name;
  name = strsave(s);
}

void possible_command::clear_name()
{
  a_delete name;
  a_delete argv;
  name = NULL;
  argv = NULL;
}

void possible_command::set_name(const char *s1, const char *s2)
{
  a_delete name;
  name = new char[strlen(s1) + strlen(s2) + 1];
  strcpy(name, s1);
  strcat(name, s2);
}

const char *possible_command::get_name()
{
  return name;
}

void possible_command::clear_args()
{
  args.clear();
}

void possible_command::append_arg(const char *s, const char *t)
{
  args += s;
  if (t)
    args += t;
  args += '\0';
}

void possible_command::insert_arg(const char *s)
{
  string str(s);
  str += '\0';
  str += args;
  args = str;
}

void possible_command::insert_args(string s)
{
  const char *p = s.contents();
  const char *end = p + s.length();
  int l = 0;
  if (p >= end)
    return;
  // find the total number of arguments in our string
  do {
    l++;
    p = strchr(p, '\0') + 1;
  } while (p < end);
  // now insert each argument preserving the order
  for (int i = l - 1; i >= 0; i--) {
    p = s.contents();
    for (int j = 0; j < i; j++)
      p = strchr(p, '\0') + 1;
    insert_arg(p);
  }
}

void possible_command::build_argv()
{
  if (argv)
    return;
  // Count the number of arguments.
  int len = args.length();
  int argc = 1;
  char *p = 0;
  if (len > 0) {
    p = &args[0];
    for (int i = 0; i < len; i++)
      if (p[i] == '\0')
	argc++;
  }
  // Build an argument vector.
  argv = new char *[argc + 1];
  argv[0] = name;
  for (int i = 1; i < argc; i++) {
    argv[i] = p;
    p = strchr(p, '\0') + 1;
  }
  argv[argc] = 0;
}

void possible_command::print(int is_last, FILE *fp)
{
  build_argv();
  if (IS_BSHELL(argv[0])
      && argv[1] != 0 && strcmp(argv[1], BSHELL_DASH_C) == 0
      && argv[2] != 0 && argv[3] == 0)
    fputs(argv[2], fp);
  else {
    fputs(argv[0], fp);
    string str;
    for (int i = 1; argv[i] != 0; i++) {
      str.clear();
      append_arg_to_string(argv[i], str);
      put_string(str, fp);
    }
  }
  if (is_last)
    putc('\n', fp);
  else
    fputs(" | ", fp);
}

void append_arg_to_string(const char *arg, string &str) // TODO quoting -> lib
{
  str += ' ';
  int needs_quoting = 0;
  int contains_single_quote = 0;
  const char*p;
  for (p = arg; *p != '\0'; p++)
    switch (*p) {
    case ';':
    case '&':
    case '(':
    case ')':
    case '|':
    case '^':
    case '<':
    case '>':
    case '\n':
    case ' ':
    case '\t':
    case '\\':
    case '"':
    case '$':
    case '?':
    case '*':
      needs_quoting = 1;
      break;
    case '\'':
      contains_single_quote = 1;
      break;
    }
  if (contains_single_quote || arg[0] == '\0') {
    str += '"';
    for (p = arg; *p != '\0'; p++)
      switch (*p) {
      case '"':
      case '\\':
      case '$':
	str += '\\';
	// fall through
      default:
	str += *p;
	break;
      }
    str += '"';
  }
  else if (needs_quoting) {
    str += '\'';
    str += arg;
    str += '\'';
  }
  else
    str += arg;
}

char **possible_command::get_argv()
{
  build_argv();
  return argv;
}

void synopsis(FILE *stream)
{
  fprintf(stream,
"Synopsis: %s [-abceghiklpstvzCENRSUVXZ] [-Fdir] [-mname] [-Tdev] [-ffam]\n"
"       [-wname] [-Wname] [-Mdir] [-dcs] [-rcn] [-nnum] [-olist] [-Parg]\n"
"       [-Darg] [-Karg] [-Larg] [-Idir] [files...]\n",
	  program_name);
}

void help()
{
  synopsis(stdout);
  fputs("\n"
"-h\tprint this message\n"
"-k\tpreprocess with preconv\n"
"-t\tpreprocess with tbl\n"
"-p\tpreprocess with pic\n"
"-e\tpreprocess with eqn\n"
"-g\tpreprocess with grn\n"
"-G\tpreprocess with grap\n"
"-s\tpreprocess with soelim\n"
"-R\tpreprocess with refer\n"
"-Tdev\tuse device dev\n"
"-mname\tread macros tmac.name\n"
"-dcs\tdefine a string c as s\n"
"-rcn\tdefine a number register c as n\n"
"-nnum\tnumber first page n\n"
"-olist\toutput only pages in list\n"
"-ffam\tuse fam as the default font family\n"
"-Fdir\tsearch dir for device directories\n"
"-Mdir\tsearch dir for macro files\n"
"-v\tprint version number\n"
"-z\tsuppress formatted output\n"
"-Z\tdon't postprocess\n"
"-a\tproduce ASCII description of output\n"
"-i\tread standard input after named input files\n"
"-wname\tenable warning name\n"
"-Wname\tinhibit warning name\n"
"-E\tinhibit all errors\n"
"-b\tprint backtraces with errors or warnings\n"
"-l\tspool the output\n"
"-c\tdisable color output\n"
"-C\tenable compatibility mode\n"
"-V\tprint commands on stdout instead of running them\n"
"-Parg\tpass arg to the postprocessor\n"
"-Larg\tpass arg to the spooler\n"
"-N\tdon't allow newlines within eqn delimiters\n"
"-S\tenable safer mode (the default)\n"
"-U\tenable unsafe mode\n"
"-Idir\tsearch dir for soelim, troff, and grops.  Implies -s\n"
"-Karg\tuse arg as input encoding.  Implies -k\n"
"-Darg\tuse arg as default input encoding.  Implies -k\n"
"\n",
	stdout);
  exit(0);
}

void usage(FILE *stream)
{
  synopsis(stream);
  fprintf(stream, "%s -h gives more help\n", program_name);
}

extern "C" {
void c_error(const char *format, const char *arg1, const char *arg2,
	     const char *arg3)
{
  error(format, arg1, arg2, arg3);
}

void c_fatal(const char *format, const char *arg1, const char *arg2,
	     const char *arg3)
{
  fatal(format, arg1, arg2, arg3);
}
} // extern "C"

// s-it2-mode
